File: react-native-nitro-player/ios/HybridEqualizer.swift
//
//  HybridEqualizer.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 04/02/26.
//

import Foundation
import NitroModules

final class HybridEqualizer: HybridEqualizerSpec {
  // MARK: - Properties

  private let core: EqualizerCore
  private var listenerIds: [(String, Int64)] = []

  // MARK: - Initialization

  override init() {
    core = EqualizerCore.shared
    super.init()
  }

  // MARK: - Enable/Disable

  func setEnabled(enabled: Bool) throws -> Promise<Void> {
    Promise.async { _ = self.core.setEnabled(enabled) }
  }

  func isEnabled() throws -> Bool {
    core.isEnabled()
  }

  // MARK: - Band Control

  func getBands() throws -> Promise<[EqualizerBand]> {
    Promise.async { self.core.getBands() }
  }

  func setBandGain(bandIndex: Double, gainDb: Double) throws -> Promise<Void> {
    Promise.async { _ = self.core.setBandGain(bandIndex: Int(bandIndex), gainDb: gainDb) }
  }

  func setAllBandGains(gains: [Double]) throws -> Promise<Void> {
    Promise.async { _ = self.core.setAllBandGains(gains) }
  }

  func getBandRange() throws -> GainRange {
    core.getBandRange()
  }

  // MARK: - Presets

  func getPresets() throws -> [EqualizerPreset] {
    core.getPresets()
  }

  func getBuiltInPresets() throws -> [EqualizerPreset] {
    core.getBuiltInPresets()
  }

  func getCustomPresets() throws -> [EqualizerPreset] {
    core.getCustomPresets()
  }

  func applyPreset(presetName: String) throws -> Promise<Void> {
    Promise.async { _ = self.core.applyPreset(presetName) }
  }

  func getCurrentPresetName() throws -> Variant_NullType_String {
    if let name = core.getCurrentPresetName() {
      return .second(name)
    }
    return .first(NullType.null)
  }

  func saveCustomPreset(name: String) throws -> Promise<Void> {
    Promise.async { _ = self.core.saveCustomPreset(name) }
  }

  func deleteCustomPreset(name: String) throws -> Promise<Void> {
    Promise.async { _ = self.core.deleteCustomPreset(name) }
  }

  // MARK: - State

  func getState() throws -> Promise<EqualizerState> {
    Promise.async { self.core.getState() }
  }

  func reset() throws -> Promise<Void> {
    Promise.async { self.core.reset() }
  }

  // MARK: - Event listeners (v2 — store IDs for cleanup)

  func onEnabledChange(callback: @escaping (_ enabled: Bool) -> Void) throws {
    let id = core.addOnEnabledChangeListener(callback)
    listenerIds.append(("onEnabledChange", id))
  }

  func onBandChange(callback: @escaping (_ bands: [EqualizerBand]) -> Void) throws {
    let id = core.addOnBandChangeListener(callback)
    listenerIds.append(("onBandChange", id))
  }

  func onPresetChange(callback: @escaping (_ presetName: Variant_NullType_String?) -> Void) throws {
    let id = core.addOnPresetChangeListener(callback)
    listenerIds.append(("onPresetChange", id))
  }

  // MARK: - Cleanup

  deinit {
    for (type, id) in listenerIds {
      switch type {
      case "onEnabledChange": _ = core.removeOnEnabledChangeListener(id: id)
      case "onBandChange":    _ = core.removeOnBandChangeListener(id: id)
      case "onPresetChange":  _ = core.removeOnPresetChangeListener(id: id)
      default: break
      }
    }
  }
}

File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/HybridEqualizer.kt
package com.margelo.nitro.nitroplayer

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.NullType
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitroplayer.equalizer.EqualizerCore

@DoNotStrip
@Keep
class HybridEqualizer : HybridEqualizerSpec() {
    private val core: EqualizerCore

    init {
        val context =
            NitroModules.applicationContext
                ?: throw IllegalStateException("React Context is not initialized")
        core = EqualizerCore.getInstance(context)
        core.ensureInitialized()
    }

    // ── Sync reads ────────────────────────────────────────────────────────────
    override fun isEnabled(): Boolean = core.isEnabled()

    override fun getBandRange(): GainRange = core.getBandRange()

    override fun getPresets(): Array<EqualizerPreset> = core.getPresets()

    override fun getBuiltInPresets(): Array<EqualizerPreset> = core.getBuiltInPresets()

    override fun getCustomPresets(): Array<EqualizerPreset> = core.getCustomPresets()

    override fun getCurrentPresetName(): Variant_NullType_String {
        val name = core.getCurrentPresetName()
        return if (name != null) {
            Variant_NullType_String.create(name)
        } else {
            Variant_NullType_String.create(NullType.NULL)
        }
    }

    // ── Async mutations (per v2 spec) ─────────────────────────────────────────
    override fun setEnabled(enabled: Boolean): Promise<Unit> = Promise.async { core.setEnabled(enabled) }

    override fun getBands(): Promise<Array<EqualizerBand>> = Promise.async { core.getBands() }

    override fun setBandGain(
        bandIndex: Double,
        gainDb: Double,
    ): Promise<Unit> = Promise.async { core.setBandGain(bandIndex.toInt(), gainDb) }

    override fun setAllBandGains(gains: DoubleArray): Promise<Unit> = Promise.async { core.setAllBandGains(gains) }

    override fun applyPreset(presetName: String): Promise<Unit> = Promise.async { core.applyPreset(presetName) }

    override fun saveCustomPreset(name: String): Promise<Unit> = Promise.async { core.saveCustomPreset(name) }

    override fun deleteCustomPreset(name: String): Promise<Unit> = Promise.async { core.deleteCustomPreset(name) }

    override fun getState(): Promise<EqualizerState> = Promise.async { core.getState() }

    override fun reset(): Promise<Unit> = Promise.async { core.reset() }

    // ── Events ────────────────────────────────────────────────────────────────
    private val listenerIds = java.util.concurrent.CopyOnWriteArrayList<Pair<String, Long>>()

    override fun onEnabledChange(callback: (enabled: Boolean) -> Unit) {
        listenerIds.add("enabled" to core.addOnEnabledChangeListener(callback))
    }

    override fun onBandChange(callback: (bands: Array<EqualizerBand>) -> Unit) {
        listenerIds.add("band" to core.addOnBandChangeListener(callback))
    }

    override fun onPresetChange(callback: (presetName: Variant_NullType_String?) -> Unit) {
        listenerIds.add("preset" to core.addOnPresetChangeListener(callback))
    }

    override fun dispose() {
        super.dispose()
        listenerIds.forEach { (type, id) ->
            when (type) {
                "enabled" -> core.removeOnEnabledChangeListener(id)
                "band" -> core.removeOnBandChangeListener(id)
                "preset" -> core.removeOnPresetChangeListener(id)
            }
        }
        listenerIds.clear()
    }
}
